daemon/service.c

```diff
@@ -0,0 +1,834 @@
+    }
+}
+
+static void
+service_dbus_name_acquired_cb(GDBusConnection *connection,
+                          const gchar *name,
+                          gpointer user_data)
+{
+    service_t *self = user_data;
+
+    if( service_dbus_name_p(name) ) {
+        log_notice("dbus name acquired");
+        service_set_nameowner(self, true);
+    }
+}
+
+static void
+service_dbus_name_lost_cb(GDBusConnection *connection,
+                      const gchar *name,
+                      gpointer user_data)
+{
+    service_t *self = user_data;
+    if( !connection ) {
+        log_err("dbus connect failed");
+        app_quit();
+    }
+    else if( service_dbus_name_p(name) ) {
+        log_err("dbus name lost");
+        service_set_nameowner(self, false);
+        app_quit();
+    }
+}
+
+static void
+service_dbus_call_cb(GDBusConnection       *connection,
+                     const gchar           *sender,
+                     const gchar           *object_path,
+                     const gchar           *interface_name,
+                     const gchar           *method_name,
+                     GVariant              *parameters,
+                     GDBusMethodInvocation *invocation,
+                     gpointer               user_data)
+{
+    service_t *self = user_data;
+
+    log_debug("from=%s object=%s method=%s.%s",
+              sender, object_path, interface_name, method_name);
+
+    auto void value_reply(GVariant *val) {
+      log_debug("reply(%p)", val);
+        g_dbus_method_invocation_return_value(invocation,
+                                              val
+                                              ? g_variant_new_tuple(&val, 1)
+                                              : NULL);
+    }
+
+    auto void error_reply(int code, const gchar *fmt, ...) {
+        va_list va;
+        va_start(va, fmt);
+        g_dbus_method_invocation_return_error_valist(invocation, G_DBUS_ERROR,
+                                                     code, fmt, va);
+        va_end(va);
+    }
+
+    if( !g_strcmp0(method_name, PERMISSIONMGR_METHOD_GET_APPLICATIONS) ) {
+        GVariantBuilder *builder = g_variant_builder_new(G_VARIANT_TYPE("as"));
+        const stringset_t *apps = applications_available(service_applications(self));
+        for( const GList *iter = stringset_list(apps); iter; iter = iter->next ) {
+            const char *appname = iter->data;
+            g_variant_builder_add(builder, "s", appname);
+        }
+        GVariant *variant = g_variant_builder_end(builder);
+        g_variant_builder_unref(builder);
+        value_reply(variant);
+    }
+    else if( !g_strcmp0(method_name, PERMISSIONMGR_METHOD_GET_APPINFO) ) {
+        const gchar *app = NULL;
+        g_variant_get(parameters, "(&s)", &app);
+        appinfo_t *appinfo = service_appinfo(self, app);
+        if( !appinfo ) {
+            error_reply(G_DBUS_ERROR_INVALID_ARGS, "Unknown application %s", app);
```

Comment:
This replies "Unknown application" and `PERMISSIONMGR_METHOD_PROMPT | PERMISSIONMGR_METHOD_QUERY` replies "Invalid application". Perhaps these should be unified.

Reply:
Done (there is way too much repetitive copy-paste code in method call handling but ... not going to address it now)